### PR TITLE
changes for cat D - pass finalisation page

### PR DESCRIPTION
**Pull request by a0dexx**
## Checklist

- [ ] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [x] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [ ] Squashed commit contains the JIRA ticket number

### CODE REVIEW

File: src/pages/page-names.constants.ts [hunk 1]
```diff
@@ -53,7 +53,7 @@ export const CAT_C = {
   DEBRIEF_PAGE: 'DebriefCatCPage',
   HEALTH_DECLARATION_PAGE: 'HealthDeclarationCatCPage',
   OFFICE_PAGE: 'OfficeCatCPage',
-  PASS_FINALISATION_PAGE: 'PassFinalisationCatCPage',
+  PASS_FINALISATION_PAGE: 'PassFinalisationCatDPage',
```
Comment (mgoldson) [inline src/pages/page-names.constants.ts]: Don't think you meant to change this one :)

File: src/pages/pass-finalisation/cat-d/_tests_/pass-finalisation.cat-d.page.spec.ts [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+    },
+    {
+      category: TestCategory.D1E,
+      code78: true,
+      transmission: TransmissionType.Manual,
+      automaticBanner: false,
+      manualBanner: false,
+      desc: 'No banner shown when manual transmission and code78 present',
+    },
+  ];
```
Comment (mgoldson) [inline src/pages/pass-finalisation/cat-d/_tests_/pass-finalisation.cat-d.page.spec.ts:184]: I'm sure the data conditioning is the same across the other categories but I would consider moving this out to separate file - maybe in a mocks folder.

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+  PassCertificateNumberChanged,
+  Code78NotPresent,
+  Code78Present,
+} from '../../../modules/tests/pass-completion/pass-completion.actions';
+import {
+  getPassCertificateNumber,
+  isProvisionalLicenseProvided,
+} from '../../../modules/tests/pass-completion/pass-completion.selector';
+import { Observable } from 'rxjs/Observable';
+import { getCandidate } from '../../../modules/tests/journal-data/cat-c/candidate/candidate.cat-c.reducer';
```
Comment (mgoldson) [inline src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts:23]: TODO needed here?

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+} from '../../../modules/tests/journal-data/common/application-reference/application-reference.reducer';
+import {
+  getApplicationNumber,
+} from '../../../modules/tests/journal-data/common/application-reference/application-reference.selector';
+import { getCurrentTest, getJournalData, getTestOutcomeText } from '../../../modules/tests/tests.selector';
+import { map } from 'rxjs/operators';
+import { Subscription } from 'rxjs/Subscription';
+import { getTests } from '../../../modules/tests/tests.reducer';
+import { PersistTests } from '../../../modules/tests/tests.actions';
+import { getVehicleDetails } from '../../../modules/tests/vehicle-details/cat-c/vehicle-details.cat-c.reducer';
```
Comment (mgoldson) [inline src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts:39]: TODO needed here?

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+import { getTestSummary } from '../../../modules/tests/test-summary/test-summary.reducer';
+import { isDebriefWitnessed, getD255 } from '../../../modules/tests/test-summary/test-summary.selector';
+import {
+  D255Yes,
+  D255No,
+  DebriefWitnessed,
+  DebriefUnwitnessed,
+} from '../../../modules/tests/test-summary/test-summary.actions';
+import { OutcomeBehaviourMapProvider } from '../../../providers/outcome-behaviour-map/outcome-behaviour-map';
+import { behaviourMap } from '../../office/office-behaviour-map.cat-c';
```
Comment (mgoldson) [inline src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts:53]: Are we using the same behaviour map as cat c?

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+import { AuthenticationProvider } from '../../../providers/authentication/authentication';
+import { BasePageComponent } from '../../../shared/classes/base-page';
+import { GearboxCategory } from '@dvsa/mes-test-schema/categories/common';
+import { PASS_CERTIFICATE_NUMBER_CTRL } from '../components/pass-certificate-number/pass-certificate-number.constants';
+import { TransmissionType } from '../../../shared/models/transmission-type';
+import { merge } from 'rxjs/observable/merge';
+import { getPassCompletion } from '../../../modules/tests/pass-completion/cat-c/pass-completion.cat-c.reducer';
+import { getCode78 } from '../../../modules/tests/pass-completion/cat-c/pass-completion.cat-c.selector';
+import { getTestCategory } from '../../../modules/tests/category/category.reducer';
+import { CategoryCode } from '@dvsa/mes-test-schema/categories/AM2';
```
Comment (mgoldson) [inline src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts:76]: Should we be getting the Category Code from common?

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+  getConductedLanguage,
+} from '../../../modules/tests/communication-preferences/communication-preferences.selector';
+import { AuthenticationProvider } from '../../../providers/authentication/authentication';
+import { BasePageComponent } from '../../../shared/classes/base-page';
+import { GearboxCategory } from '@dvsa/mes-test-schema/categories/common';
+import { PASS_CERTIFICATE_NUMBER_CTRL } from '../components/pass-certificate-number/pass-certificate-number.constants';
+import { TransmissionType } from '../../../shared/models/transmission-type';
+import { merge } from 'rxjs/observable/merge';
+import { getPassCompletion } from '../../../modules/tests/pass-completion/cat-c/pass-completion.cat-c.reducer';
+import { getCode78 } from '../../../modules/tests/pass-completion/cat-c/pass-completion.cat-c.selector';
```
Comment (mgoldson) [inline src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts]: TODOs?

**Review by mgoldson — APPROVED** [review]
A few minor things to look at but looks good. A description and screenshot in the PR would be nice.

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+  d255$: Observable<boolean>;
+  debriefWitnessed$: Observable<boolean>;
+  conductedLanguage$: Observable<string>;
+  code78$: Observable<boolean>;
+  testCategory$: Observable<CategoryCode>;
+}
+
+@IonicPage()
+@Component({
+  selector: '.pass-finalisation-cat-c-page',
```
Comment (matthew2564) [inline src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts]: Update selector to  '.pass-finalisation-cat-d-page'

File: src/pages/pass-finalisation/cat-d/_tests_/pass-finalisation.cat-d.page.spec.ts [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+import { TransmissionComponent } from '../../../../components/common/transmission/transmission';
+import { FormGroup, FormControl, Validators } from '@angular/forms';
+import { PASS_CERTIFICATE_NUMBER_CTRL }
+  from '../../components/pass-certificate-number/pass-certificate-number.constants';
+import { Code78Component } from '../components/code-78/code-78';
+import { TransmissionType } from '../../../../shared/models/transmission-type';
+import { configureTestSuite } from 'ng-bullet';
+import { TestCategory } from '@dvsa/mes-test-schema/category-definitions/common/test-category';
+
+describe('PassFinalisationCatCPage', () => {
```
Comment (matthew2564) [inline src/pages/pass-finalisation/cat-d/_tests_/pass-finalisation.cat-d.page.spec.ts]: Update describe to 'PassFinalisationCatDPage'

File: src/pages/page-names.constants.ts (hunk 1 above)
Comment (amobarak) [reply src/pages/page-names.constants.ts]: Yeah, this would break the app

Comment (a0dexx) [reply src/pages/page-names.constants.ts]: done

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts (hunk 3 above)
Comment (a0dexx) [reply src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts:23]: done

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts (hunk 4 above)
Comment (a0dexx) [reply src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts:39]: done

File: src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+  Code78NotPresent,
+  Code78Present,
+} from '../../../modules/tests/pass-completion/pass-completion.actions';
+import {
+  getPassCertificateNumber,
+  isProvisionalLicenseProvided,
+} from '../../../modules/tests/pass-completion/pass-completion.selector';
+import { Observable } from 'rxjs/Observable';
+// TODO: create cat d reducer
+import { getCandidate } from '../../../modules/tests/journal-data/cat-c/candidate/candidate.cat-c.reducer';
```
Comment (rossfauldsbjss) [inline src/pages/pass-finalisation/cat-d/pass-finalisation.cat-d.page.ts:23]: i've created cat d reducers for this and merged in. You could use this now.

**Review by appleabs — APPROVED** [review]

**Pull request merged**